### ISSUE TITLE
Define each event type is composed or not, update the spec

### ISSUE THREAD

**Issue by TakayoshiKochi** [w3c/clipboard-apis#61]
It looks current ClipboardEvent doesn't define anything about whether each event is composed or not.

Once it was discussed at https://github.com/w3c/webcomponents/issues/513#issuecomment-224468080) but no one has ever revisited this.

/cc @annevk @hayatoito @rniwa

**Comment by TakayoshiKochi** [comment]
The current status is that, as the spec says nothing about `composed` flag, but `ClipboardEventInit` dictionary inherits `EventInit`, `composed` flag is false by default.

Considering the similarity with UI events, shall we flip the default to be true?

**Comment by hayatoito** [comment]
@TakayoshiKochi  What's your opinion?

**Comment by TakayoshiKochi** [comment]
My opinion is composed=true by default.

**Comment by annevk** [comment]
For these events, when they're not created synthetically? If so, seems fine to me and matches other UI events.

**Comment by TakayoshiKochi** [comment]
@annevk yes, so follow other UI events.

I found @rniwa already commented: https://github.com/w3c/webcomponents/issues/513#issuecomment-230989199
> we should make `copy`, `paste`, `cut` and their `before*` variants composed since they don't have any node leakage problem, and yet are very important user gestures.

So I'll start preparing a PR for the clipboard spec.

**Comment by TakayoshiKochi** [comment]
Did not find any MS people in the original thead https://github.com/w3c/webcomponents/issues/513, but cc @travisleithead - sorry I missed you.

**Comment by hayatoito** [comment]
Could you help me to understand the motivation of making clipboard events composed one?

Given that users can't access a node where an event is fired, what is the use case for that?

**Comment by TakayoshiKochi** [comment]
WebKit implements clipboard events as composed=true by default, while Blink implements them composed=false.  So we'd like to clarify which is the default.

I thought use cases are obvious - like catching the timing when copy or paste operation happens within shadow root, observe it from outside, and do something against it.  For example, if you want to use an editor component with shadow root, the embedder of the component may want to observe copy/paste events.  Even without looking at the real target, you can access ClipboardData via the event.

**Comment by hayatoito** [comment]
Thanks, however, the explanation is still lacking why a light tree want to know copy & paste events which is happening in a shadow tree. What is the difference between clipboard events and other events?

I meant your explanation is generic and might be applied to all kinds of events which are happening in a shadow tree; "A light tree want to observe what is happening in a shadow tree".

Also, I have a concern about encapsulation. It would be bad that a light tree can know the content of a shadow tree via the event's ClipboardData, if we make clipboard events composed=true.

**Comment by TakayoshiKochi** [comment]
What's the difference with `InputEvent` (which is defined composed=true) and `ClipboardEvent`, then?

I don't think on `copy`/`cut` you can read the to-be-placed-in-clipboard data from ClipboardData, but just you can set your own arbitrary data to it.  But I may miss something.

**Comment by hayatoito** [comment]
Thanks. Given your explanation so far, I guess your motivation came from:

- ClipboardEvents and InputEvents are similar. So make ClipboardEvents composed=true for consistency.

Is the consistency the only motivation?  Can I assume that we have not heard any use case or request from customers, like Polymer, to make this happen?

> I don't think on copy/cut you can read the to-be-placed-in-clipboard data from ClipboardData, but just you can set your own arbitrary data to it. But I may miss something.

I don't know neither. Please make sure there wouldn't be an unintentional content leak in CblipboardEvent, before making PR.

**Comment by TakayoshiKochi** [comment]
I don't know this is a meaningful use case, but I found this while trying to convert `<input>` `<textarea>` elements to use V1 user-agent shadow DOM. Unless ClipboardEvent is composed, any clipboard event happened in the internal user agent shadow doesn't propagate outside the UA shadow.
So it means you cannot create your own input or textarea custom element, although technically it is possible to redispatch an event when a clipboard event is caught on the boundary shadow root.  So the request is coming from myself, so to say.

In the Shadow DOM V0 era, such clipboard events are not listed in the "scoped events" list, and crossed the shadow boundary already (including user-agent shadow boundary).

I think I made sure as much as I can to check if anything leaks, but no one can be perfect.
I'm asking people cc'ed here, and clipboard API authors for sanity checking and opinions.

**Comment by hayatoito** [comment]
Thanks. I'm happy to know from where your motivation came.

What I can say now is that our implementation detail shouldn't be a reason to change ClipboardEvent's web facing behavior.

I don't have any strong opinion whether we should make ClipboardEvent composed=true or not. The consistency might matter here, but I am not sure.

It would be really nice that we could find any use case or feature request from a web developer to make a decision.

**Comment by annevk** [comment]
@hayatoito you don't agree that our consideration to date has been "is it a type of a UI event"? I thought that our principle was that we'd make UI events composed, unless they'd lead to leaks, in which case we need to do more work to figure out what the best solution is.

**Comment by TakayoshiKochi** [comment]
@kevinpschaaf do you have any experience / opinion from Polymer?

**Comment by hayatoito** [comment]
> "is it a type of a UI event"? I thought that our principle was that we'd make UI events composed.

Yup, I don't object to this. I remember that is our principle, which we should honor.

If we think ClibboardEvent is a type of a UI event, we might want to make ClipboardEvent inherit UIEvents *officially* too, as well as making it composed..

```
interface ClipboardEvent : UIEvent {
```

I think this would be possible.

**Comment by travisleithead** [comment]
I don't immediately see a specific scenario for `composed=true`, but consistency with the other UIEvents is compelling enough to have that be the default IMO. @gked

**Comment by kevinpschaaf** [comment]
@TakayoshiKochi I like the consistency with UI events, but also don't have a compelling use case.  Maybe @justinfagnani can think of a concrete use case from his work on a wysiwyg designer.

**Comment by justinfagnani** [comment]
I don't have anything off the top of my head. My previous WYSIWYG work already intercepted all events, and editing in general, and so could get clipboard events it needed. The move to contenteditable might change this, but I don't see where we couldn't use capture to achieve the same.

Seems like consistency is as good of a reason as any for composed.

**Comment by TakayoshiKochi** [comment]
Thanks for your comments - all seems no objection, while no strong use case yet.

@garykac do you have any opinion whether ClipboardEvent should or should not inherit from UIEvent?

**Comment by rniwa** [comment]
Making `copy`, `paste`, `cut`, and their before* variant composed makes sense to me.

**Comment by garykac** [comment]
Having ClipboardEvent inherit from UIEvent (instead of Event) seems reasonable to me.

I notice that UIEvent doesn't mention anything about "composed" in the spec (other than having a "composed" entry in the table for most events). It feels like there should be something mentioned there if we consider UIEvents to be containers for it.  Is there a definition that I can link to from the spec?

**Comment by annevk** [comment]
No, all events can be composed, it's just that most UI events are so by default. I don't think we should make that mean "has to inherit from UIEvent" though. Both its `view` and `detail` members are rather useless and I don't see why we'd add those to newly defined UI events.

**Comment by TakayoshiKochi** [comment]
@garykac thanks for the comment.
The conclusion at https://github.com/w3c/webcomponents/issues/513#issuecomment-224486249 was to iterate on each UA-firing events (not UI events) to see if it is a good fit for having `composed=true` by default or not.  So not all UI events are `composed=true` by default.  (while I'm writing this @annevk outpaced me).

@annevk I agree that I don't see much sense ClipBoardEvents having `view` or `detail` - we don't make clipboard events inherit UIEvent just because we want to make its `composed` default to be true. So to answer @hayatoito's question https://github.com/w3c/clipboard-apis/issues/61#issuecomment-357930616 we don't have to make clipboard events inherit from UI Events, but we can do it independently from `composed` default value change, if necessary.

Let's close the loop for whether or not inherit from UIEvent.

**Comment by annevk** [comment]
I'd prefer not inheriting from `UIEvent`, unless someone can show the value of doing so.

**Comment by hayatoito** [comment]
I see. It sounds that we should discourage the newly defined *UI-related* Event to inherit `UIEvent` because its `view` and `detail` members are useless there.

I am afraid that someone would think that there is an inconsistency between `ClipboardEvent` and existing UI Events. It would be better to add a note to the spec like:
"For newly defined UI events, we don't recommend them to inherit UI events because ....". That would be helpful.

**Comment by TakayoshiKochi** [comment]
I filed a separate PR for UIEvent spec for adding a note about inheritance.

So we all agree we don't inherit UIEvent for ClipboadEvent, then are we still blocked on compelling real-world use case to be `composed=true` by default? Or on the contrary, did Blink have any problem ClipboardEvents have not been scoped for Shadow DOM V0?

**Comment by hayatoito** [comment]
I think we can proceed without having a real world use case.

**Comment by TakayoshiKochi** [comment]
Thanks, @hayatoito.

@garykac, could you review and merge #62 ?
